mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java

```diff
@@ -0,0 +1,80 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.postgres.mail.dao;
+
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.HASH_BASE_SUBJECT;
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.HASH_MIME_MESSAGE_ID;
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.MESSAGE_ID;
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.TABLE_NAME;
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.THREAD_ID;
+import static org.apache.james.mailbox.postgres.mail.dao.PostgresThreadModule.PostgresThreadTable.USERNAME;
+
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.ThreadId;
+import org.apache.james.mailbox.postgres.PostgresMessageId;
+import org.jooq.Record;
+
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresThreadDAO {
+    private final PostgresExecutor postgresExecutor;
+
+    public PostgresThreadDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Mono<Void> insertSome(Username username, Set<Integer> hashMimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Integer> hashBaseSubject) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.batch(
+            hashMimeMessageIds.stream().map(hashMimeMessageId -> dslContext.insertInto(TABLE_NAME)
+                .set(USERNAME, username.asString())
+                .set(HASH_MIME_MESSAGE_ID, hashMimeMessageId)
+                .set(MESSAGE_ID, ((PostgresMessageId) messageId).asUuid())
+                .set(THREAD_ID, ((PostgresMessageId) threadId.getBaseMessageId()).asUuid())
+                .set(HASH_BASE_SUBJECT, hashBaseSubject.orElse(null)))
+                .collect(ImmutableList.toImmutableList()))));
+    }
+
+    public Flux<Pair<Optional<Integer>, ThreadId>> findThreads(Username username, Set<Integer> hashMimeMessageIds) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(THREAD_ID, HASH_BASE_SUBJECT)
+            .from(TABLE_NAME)
+            .where(USERNAME.eq(username.asString()))
+            .and(HASH_MIME_MESSAGE_ID.in(hashMimeMessageIds))))
```

Comment:
Can I break this by supplying a message with an absurdly large number of references?

Reply:
How many?

Reply:
Maybe we can bathc them by group of 16?

Reply:
It seems that Postgres is still able to handle hundreds elements in IN clause without any issues. What is maximum length of hashMimeMessageIds?

Reply:
> What is maximum length of hashMimeMessageIds?

Can not say for sure. If the conversation keeps being looped then the length could be very big :D

On IN limitation, I remember @vttranlina apply it somewhere in the mailbox component?

Reply:
partition added